File: backend/services/stigmer-server/pkg/encryption/keymanager.go
package encryption

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
)

const (
	// EnvKeyName is the environment variable name for the encryption key
	EnvKeyName = "STIGMER_ENCRYPTION_KEY"

	// KeyFileName is the default filename for the local key file
	KeyFileName = "encryption.key"

	// KeyFileDir is the directory under home for storing the key file
	KeyFileDir = ".stigmer"

	// KeyFilePermissions are the permissions for the key file (owner read/write only)
	KeyFilePermissions = 0600

	// KeyDirPermissions are the permissions for the key directory
	KeyDirPermissions = 0700
)

// GetOrCreateKey loads the encryption key with the following priority:
//
//  1. STIGMER_ENCRYPTION_KEY environment variable (Base64-encoded 32-byte key)
//  2. ~/.stigmer/encryption.key file (raw 32-byte key)
//  3. Auto-generate and persist to file if neither exists
//
// Returns nil if encryption should be disabled (no key available and
// auto-generation is not appropriate for the environment).
func GetOrCreateKey() ([]byte, error) {
	return GetOrCreateNamedKey(EnvKeyName, KeyFileName)
}

// GetOrCreateNamedKey is GetOrCreateKey generalized over the env var and key
// file name, so sibling key material (the runner-token signing key, oss#535)
// rides the same env-var -> ~/.stigmer file -> auto-generate convention
// instead of growing a divergent loader.
func GetOrCreateNamedKey(envVar, fileName string) ([]byte, error) {
	// 1. Check environment variable (highest priority)
	if envKey := os.Getenv(envVar); envKey != "" {
		key, err := base64.StdEncoding.DecodeString(envKey)
		if err != nil {
			return nil, fmt.Errorf("invalid Base64 encoding in %s: %w", envVar, err)
		}
		if len(key) != KeySize {
			return nil, fmt.Errorf("%s must be exactly 32 bytes (256 bits) when decoded, got %d bytes", envVar, len(key))
		}
		return key, nil
	}

	// 2. Check local file
	keyPath, err := getNamedKeyFilePath(fileName)
	if err != nil {
		return nil, err
	}

	if key, err := loadKeyFromFile(keyPath); err == nil {
		return key, nil
	}

	// 3. Auto-generate new key (for local development)
	// In production, the key should be configured via environment variable
	key, err := generateKey()
	if err != nil {
		return nil, fmt.Errorf("failed to generate key for %s: %w", fileName, err)
	}

	// Persist for future use
	if err := saveKeyToFile(keyPath, key); err != nil {
		// Log warning but don't fail - the key is still usable
		fmt.Fprintf(os.Stderr, "Warning: could not save key to %s: %v\n", keyPath, err)
	}

	return key, nil
}

// GetKey loads the encryption key without auto-generation.
// Returns nil if no key is configured.
// Use this in production where auto-generation is not desired.
func GetKey() ([]byte, error) {
	// 1. Check environment variable
	if envKey := os.Getenv(EnvKeyName); envKey != "" {
		key, err := base64.StdEncoding.DecodeString(envKey)
		if err != nil {
			return nil, fmt.Errorf("invalid Base64 encoding in %s: %w", EnvKeyName, err)
		}
		if len(key) != KeySize {
			return nil, fmt.Errorf("%s must be exactly 32 bytes (256 bits) when decoded, got %d bytes", EnvKeyName, len(key))
		}
		return key, nil
	}

	// 2. Check local file
	keyPath, err := getKeyFilePath()
	if err != nil {
		return nil, nil // No key file path available
	}

	key, err := loadKeyFromFile(keyPath)
	if err != nil {
		return nil, nil // No key file exists
	}

	return key, nil
}

// getKeyFilePath returns the path to the local key file (~/.stigmer/encryption.key)
func getKeyFilePath() (string, error) {
	return getNamedKeyFilePath(KeyFileName)
}

// getNamedKeyFilePath returns the path to a named key file under ~/.stigmer.
func getNamedKeyFilePath(fileName string) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("could not determine home directory: %w", err)
	}
	return filepath.Join(home, KeyFileDir, fileName), nil
}

// loadKeyFromFile reads the raw key bytes from a file
func loadKeyFromFile(path string) ([]byte, error) {
	// Check file exists
	info, err := os.Stat(path)
	if err != nil {
		return nil, err
	}

	// Verify secure permissions (owner read/write only)
	mode := info.Mode().Perm()
	if mode != KeyFilePermissions {
		return nil, fmt.Errorf("key file %s has insecure permissions %o, expected %o", path, mode, KeyFilePermissions)
	}

	// Read key
	key, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	// Validate size
	if len(key) != KeySize {
		return nil, fmt.Errorf("key file %s has invalid size: expected %d bytes, got %d", path, KeySize, len(key))
	}

	return key, nil
}

// saveKeyToFile writes the raw key bytes to a file with secure permissions
func saveKeyToFile(path string, key []byte) error {
	// Ensure directory exists with secure permissions
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, KeyDirPermissions); err != nil {
		return fmt.Errorf("failed to create key directory: %w", err)
	}

	// Write key with secure permissions
	if err := os.WriteFile(path, key, KeyFilePermissions); err != nil {
		return fmt.Errorf("failed to write key file: %w", err)
	}

	return nil
}

// generateKey generates a cryptographically secure random 32-byte key
func generateKey() ([]byte, error) {
	key := make([]byte, KeySize)
	if _, err := rand.Read(key); err != nil {
		return nil, err
	}
	return key, nil
}

// GenerateKeyBase64 generates a new random encryption key and returns it
// as a Base64-encoded string suitable for use in environment variables.
// This is a helper for key generation - the key is not stored.
func GenerateKeyBase64() (string, error) {
	key, err := generateKey()
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(key), nil
}

// ValidateKeyBase64 validates that a Base64-encoded string represents
// a valid 32-byte encryption key.
func ValidateKeyBase64(encoded string) error {
	key, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return fmt.Errorf("invalid Base64 encoding: %w", err)
	}
	if len(key) != KeySize {
		return fmt.Errorf("key must be exactly 32 bytes (256 bits), got %d bytes", len(key))
	}
	return nil
}

File: backend/services/stigmer-server/src/encryption/key-manager.ts
/**
 * Named-key loader — ports pkg/encryption/keymanager.go.
 *
 * One convention for every key the server holds: env var (Base64) →
 * ~/.stigmer/<file> (0600) → auto-generate and persist. The
 * loader is generalized over the env var, the file name and a KeyCodec so
 * sibling key material rides the same ladder instead of growing a
 * divergent loader: the runner-token signing key (oss#535) and the
 * encryption key are 32 raw bytes (RAW_32_BYTE_KEY); the platform-token
 * signing key is an RSA private key (platformtoken/key-ring.ts). The codec
 * says what a key IS — how an env value and a file decode, how one is
 * generated and written — and nothing about the ladder, which stays one.
 *
 * Ladder semantics, exactly Go's, for every codec:
 *   - An EXPLICITLY configured env key that is unusable (bad Base64, wrong
 *     length, not the key type the codec holds) is an ERROR, never a
 *     degrade — silently ignoring deliberate configuration would be worse
 *     than refusing to boot.
 *   - A key file that fails its load checks (permissions other than 0600,
 *     content the codec does not accept) is never adopted; the ladder
 *     FALLS THROUGH to auto-generate, which overwrites the file's content —
 *     Go's shipped behavior (loadKeyFromFile error → generate → save),
 *     ported as-is.
 *   - Auto-generation persists for future boots; a persist FAILURE is a
 *     stderr warning only — the key is still usable for this process.
 *
 * `env` and `homeDir` are injectable with process defaults — the
 * loadConfig(env = process.env) idiom (boot/config.ts) — so unit tests
 * stay hermetic and never touch the real ~/.stigmer (DD-002). The key env
 * vars deliberately do NOT ride ServerConfig: Go's pkg/config never sees
 * them either, and config.ts's contract is that no entry exists before
 * the code that reads it.
 */
import { randomBytes } from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

/** AES-256 / HMAC-SHA256 key size (32 bytes = 256 bits) — Go KeySize. */
export const KEY_SIZE = 32;

/** Directory under home for key files — Go KeyFileDir. */
export const KEY_FILE_DIR = ".stigmer";

/** Key file permissions (owner read/write only) — Go KeyFilePermissions. */
export const KEY_FILE_PERMISSIONS = 0o600;

/** Key directory permissions — Go KeyDirPermissions. */
export const KEY_DIR_PERMISSIONS = 0o700;

export interface KeyLoaderOptions {
  /** Environment map; defaults to the live process env. */
  readonly env?: NodeJS.ProcessEnv;
  /** Home directory for ~/.stigmer; defaults to os.homedir(). */
  readonly homeDir?: string;
}

/**
 * What one kind of key is, for the ladder. `fromEnv` receives the env
 * value already strictly Base64-decoded and THROWS when the bytes are not
 * a usable key (explicit configuration is never silently ignored);
 * `fromFile` answers `undefined` for content it does not accept, so the
 * ladder falls through exactly as it does for a wrong-sized raw key.
 */
export interface KeyCodec<Key> {
  fromEnv(envVar: string, decoded: Buffer): Key;
  fromFile(content: Buffer): Key | undefined;
  generate(): Key;
  toFile(key: Key): Buffer;
}

/** The 32 raw bytes the encryption and runner-token keys are. */
export const RAW_32_BYTE_KEY: KeyCodec<Buffer> = {
  fromEnv(envVar, decoded) {
    if (decoded.length !== KEY_SIZE) {
      throw new Error(
        `${envVar} must be exactly 32 bytes (256 bits) when decoded, got ${decoded.length} bytes`,
      );
    }
    return decoded;
  },
  fromFile(content) {
    return content.length === KEY_SIZE ? content : undefined;
  },
  generate() {
    return randomBytes(KEY_SIZE);
  },
  toFile(key) {
    return key;
  },
};

/**
 * Go GetOrCreateNamedKey: env var → key file → auto-generate, for a 32-byte
 * raw key. Throws only on unusable EXPLICIT configuration (bad env value)
 * or when no key can be produced at all.
 */
export function getOrCreateNamedKey(
  envVar: string,
  fileName: string,
  options: KeyLoaderOptions = {},
): Buffer {
  return getOrCreateKey(RAW_32_BYTE_KEY, envVar, fileName, options);
}

/** The ladder over any codec: env var → key file → auto-generate. */
export function getOrCreateKey<Key>(
  codec: KeyCodec<Key>,
  envVar: string,
  fileName: string,
  options: KeyLoaderOptions = {},
): Key {
  const env = options.env ?? process.env;

  // 1. Environment variable (highest priority) — Base64-encoded.
  const envValue = env[envVar];
  if (envValue !== undefined && envValue !== "") {
    return codec.fromEnv(envVar, decodeEnvBase64(envVar, envValue));
  }

  // 2. Local key file — refused on insecure permissions or content the
  //    codec does not accept.
  const keyPath = namedKeyFilePath(fileName, options);
  const content = readKeyFile(keyPath);
  const fromFile = content === undefined ? undefined : codec.fromFile(content);
  if (fromFile !== undefined) {
    return fromFile;
  }

  // 3. Auto-generate (local development); persist for future boots.
  const key = codec.generate();
  try {
    saveKeyToFile(keyPath, codec.toFile(key));
  } catch (error) {
    // Warn but don't fail — the key is still usable (Go's posture).
    process.stderr.write(
      `Warning: could not save key to ${keyPath}: ${error instanceof Error ? error.message : String(error)}\n`,
    );
  }
  return key;
}

/** Strict Base64 decode for an env-configured key. */
function decodeEnvBase64(envVar: string, value: string): Buffer {
  const decoded = Buffer.from(value, "base64");
  // Node's base64 decoder is lenient (skips invalid characters, accepts
  // missing padding) and never throws; Go's StdEncoding errors on any
  // non-canonical input. Round-tripping — Buffer always re-emits the
  // canonical padded form — restores Go's strictness.
  if (decoded.toString("base64") !== value) {
    throw new Error(`invalid Base64 encoding in ${envVar}`);
  }
  return decoded;
}

/** Path to a named key file under <home>/.stigmer. */
export function namedKeyFilePath(
  fileName: string,
  options: KeyLoaderOptions = {},
): string {
  const home = options.homeDir ?? os.homedir();
  return path.join(home, KEY_FILE_DIR, fileName);
}

/**
 * Reads a key file's bytes, mirroring Go loadKeyFromFile: missing file →
 * undefined (fall through the ladder); present with insecure permissions
 * also falls through — exactly Go's behavior, where any load error falls
 * to auto-generate. The 0600 check is skipped on Windows, where POSIX
 * modes are not meaningful.
 */
function readKeyFile(keyPath: string): Buffer | undefined {
  let stat: fs.Stats;
  try {
    stat = fs.statSync(keyPath);
  } catch {
    return undefined;
  }
  if (process.platform !== "win32") {
    const mode = stat.mode & 0o777;
    if (mode !== KEY_FILE_PERMISSIONS) {
      return undefined;
    }
  }
  try {
    return fs.readFileSync(keyPath);
  } catch {
    return undefined;
  }
}

/** Writes the key with secure permissions (dir 0700, file 0600). */
function saveKeyToFile(keyPath: string, content: Buffer): void {
  fs.mkdirSync(path.dirname(keyPath), {
    recursive: true,
    mode: KEY_DIR_PERMISSIONS,
  });
  fs.writeFileSync(keyPath, content, { mode: KEY_FILE_PERMISSIONS });
}
